### PR TITLE
Conversational dataset support for `DPOTrainer`

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by kashif — APPROVED** [review]

File: trl/data_utils.py [hunk 1]
```diff
@@ -280,15 +280,17 @@ def maybe_unpair_preference_dataset(dataset: DatasetType, num_proc: Optional[int
...
     Extracts the shared prompt from a preference data example, where the prompt is implicit within both
     the chosen and rejected completions.
 
     For more details, see [`maybe_extract_prompt`].
     """
     for idx in range(min(len(example["chosen"]), len(example["rejected"]))):
-        if example["chosen"][idx]["content"] != example["rejected"][idx]["content"]:
+        if example["chosen"][idx] != example["rejected"][idx]:
+            if example["chosen"][idx - 1] == " ":  # remove space before the prompt
+                idx -= 1
```
Comment (qgallouedec) [inline trl/data_utils.py:293]:
```python
str1 = "I am Quentin"
str2 = "I am in Lyon"
# What we want:
prompt = "I am"
# What we don't want:
prompt = "I am "
```


That's why, when the prompt ends with a space, we take `idx-1` instead.

File: docs/source/dataset_formats.mdx [hunk 2]
```diff
@@ -180,6 +180,8 @@ preference_example = {"prompt": "The sky is", "chosen": " blue.", "rejected": "
 preference_example = {"chosen": "The sky is blue.", "rejected": "The sky is green."}
 ```
 
+Some preference datasets can be found with [the tag `dpo` on Hugging Face Hub](https://huggingface.co/datasets?other=dpo). You can also explore the [librarian-bots' DPO Collections](https://huggingface.co/collections/librarian-bots/direct-preference-optimization-datasets-66964b12835f46289b6ef2fc) to identify preference datasets.
```
Comment (lewtun) [inline docs/source/dataset_formats.mdx:183]: Not for this PR, but WDYT about putting these Hub filters as a badge at the top of each trainer doc, similar to how `transformers` does it for architectures: https://huggingface.co/docs/transformers/model_doc/bert

Screenshot 2024-10-01 at 17 31 20

You can generate these badges with: https://shields.io

File: docs/source/dpo_trainer.mdx [hunk 3]
```diff
@@ -1,166 +1,126 @@
...
+from trl import DPOConfig, DPOTrainer
+from transformers import AutoModelForCausalLM, AutoTokenizer
 
-The DPO trainer expects a model of `AutoModelForCausalLM` or `AutoModelForVision2Seq`, compared to PPO that expects `AutoModelForCausalLMWithValueHead` for the value function.
+model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
+tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
+train_dataset = load_dataset("trl-lib/Capybara-Preferences", split="train")
 
-## Using the `DPOTrainer`
+training_args = DPOConfig(output_dir="dpo-qwen2", logging_steps=10)
```
Comment (lewtun) [inline docs/source/dpo_trainer.mdx:47]: turbo nit:
```suggestion
training_args = DPOConfig(output_dir="Qwen2-0.5B-DPO", logging_steps=10)
```

File: docs/source/dpo_trainer.mdx [hunk 4]
```diff
@@ -1,166 +1,126 @@
...
-After this one can then call:
+Distributed across 8 GPUs, the training takes approximately 3 minutes. You can verify the training progress by checking the reward graph. An increasing trend in the reward margin indicates that the model is improving and generating better responses over time.
 
-```py
-dpo_trainer.train()
-```
+![](https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/dpo-qwen2-reward-margin.png)
 
-Note that the `beta` is the temperature parameter for the DPO loss, typically something in the range of `0.1` to `0.5`. We ignore the reference model as `beta` -> 0.
+To see how the [trained model](https://huggingface.co/trl-lib/dpo-qwen2) performs, use the following code to generate completions:
```
Comment (lewtun) [inline docs/source/dpo_trainer.mdx:62]: WDYT about using `trl chat` here as a nice demo to show users it exists?

File: docs/source/dpo_trainer.mdx [hunk 5]
```diff
@@ -169,59 +129,71 @@ While training and evaluating we record the following reward metrics:
 - `rewards/accuracies`: mean of how often the chosen rewards are > than the corresponding rejected rewards
 - `rewards/margins`: the mean difference between the chosen and corresponding rejected rewards
 
+## Loss functions
+
+The DPO algorithm supports several loss functions. The loss function can be set using the `loss_type` parameter in the [`DPOConfig`]. The following loss functions are supported:
```
Comment (lewtun) [inline docs/source/dpo_trainer.mdx:134]: Amazing doc!

File: trl/trainer/dpo_config.py [hunk 6]
```diff
@@ -130,6 +133,7 @@ class DPOConfig(TrainingArguments):
             DPO loss. The paper recommends `rpo_alpha=1.0`.
     """
 
+    learning_rate: float = 1e-6
```
Comment (lewtun) [inline trl/trainer/dpo_config.py:136]: Great idea to set a good default value!

**Review by lewtun — APPROVED** [review]
Really nice clean up of the DPO preprocessing and docs 🔥. Great stuff @qgallouedec !

I know we don't currently have any proper regression tests, but WDYT about running the `dpo.py` script on `main` and this branch to sanity check the margins / loss look ok before merging?

File: docs/source/dataset_formats.mdx (hunk 2 above)
Comment (qgallouedec) [reply docs/source/dataset_formats.mdx:183]: #2149

File: docs/source/dpo_trainer.mdx (hunk 3 above)
Comment (qgallouedec) [reply docs/source/dpo_trainer.mdx]: Model on hub renamed as well

File: docs/source/dpo_trainer.mdx (hunk 4 above)
Comment (qgallouedec) [reply docs/source/dpo_trainer.mdx]: Something like this?

image

File: docs/source/dpo_trainer.mdx [hunk 7]
```diff
@@ -169,59 +130,71 @@ While training and evaluating we record the following reward metrics:
...
+| `"bco_pair"`                           | The [BCO](https://huggingface.co/papers/2404.04656) authors train a binary classifier whose logit serves as a reward so that the classifier maps {prompt, chosen completion} pairs to 1 and {prompt, rejected completion} pairs to 0.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
```
Comment (kashif) [inline docs/source/dpo_trainer.mdx:145]: shall we also say that the actual BCO Trainer is in the `BCOTrainer`?

Comment (qgallouedec) [reply docs/source/dpo_trainer.mdx:145]: Good point. Are they equivalent? Why do we have both?

Comment (kashif) [reply docs/source/dpo_trainer.mdx:145]: so the one in the DPO is the "paired" version of BCO... in theory it can be removed... BCO is like KTO to a certain extent

Comment (qgallouedec) [reply docs/source/dpo_trainer.mdx:145]: How would you formulate it for the doc?

Comment (kashif) [inline docs/source/dpo_trainer.mdx:145]:
```suggestion
| `"bco_pair"`                           | The [BCO](https://huggingface.co/papers/2404.04656) authors train a binary classifier whose logit serves as a reward so that the classifier maps {prompt, chosen completion} pairs to 1 and {prompt, rejected completion} pairs to 0.  For un-paried data use the [`BCOTrainer`].                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
```

**Comment by qgallouedec** [comment]
```
accelerate launch examples/scripts/dpo.py \
    --dataset_name trl-lib/ultrafeedback_binarized \
    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
    --learning_rate 5.0e-7 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 8 \
    --gradient_checkpointing \
    --logging_steps 25 \
    --eval_strategy steps \
    --eval_steps 50 \
    --output_dir Qwen2-0.5B-DPO-main \
    --no_remove_unused_columns
```

Screenshot 2024-10-02 at 09 41 25

**Review by kashif — APPROVED** [review]
